### PR TITLE
Evaluate defaults during async insert safer

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 76766a200e480fa746f6d93bad69452a2550fb61 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55253/76766a200e480fa746f6d93bad69452a2550fb61/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6416587089">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55253/76766a200e480fa746f6d93bad69452a2550fb61/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/AsynchronousInsertQueue.cpp [hunk 1]
```diff
@@ -60,10 +60,12 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
-AsynchronousInsertQueue::InsertQuery::InsertQuery(const ASTPtr & query_, const Settings & settings_)
+AsynchronousInsertQueue::InsertQuery::InsertQuery(const ASTPtr & query_, const Settings & settings_, const std::optional<UUID> & user_id_, const std::vector<UUID> & current_roles_)
     : query(query_->clone())
     , query_str(queryToString(query))
     , settings(settings_)
+    , user_id(user_id_)
+    , current_roles(current_roles_)
```
Comment (CurtizJ) [inline src/Interpreters/AsynchronousInsertQueue.cpp:68]: Those fields should be added to hash and `==` operator.

Comment (vitlibar) [reply src/Interpreters/AsynchronousInsertQueue.cpp:68]: done

**Review by CurtizJ — APPROVED** [review]

File: tests/queries/0_stateless/02247_written_bytes_quota.sh [hunk 2]
```diff
@@ -16,10 +16,13 @@ ${CLICKHOUSE_CLIENT} -q "CREATE ROLE r02247"
 ${CLICKHOUSE_CLIENT} -q "CREATE USER u02247"
 ${CLICKHOUSE_CLIENT} -q "GRANT ALL ON *.* TO r02247"
 ${CLICKHOUSE_CLIENT} -q "GRANT r02247 to u02247"
-${CLICKHOUSE_CLIENT} -q "CREATE QUOTA q02247 FOR INTERVAL 100 YEAR MAX WRITTEN BYTES = 25 TO r02247"
+${CLICKHOUSE_CLIENT} -q "CREATE QUOTA q02247 FOR INTERVAL 100 YEAR MAX WRITTEN BYTES = 30 TO r02247"
```
Comment (CurtizJ) [inline tests/queries/0_stateless/02247_written_bytes_quota.sh:19]: It looks that quota is accounted twice for user now.

Comment (CurtizJ) [reply tests/queries/0_stateless/02247_written_bytes_quota.sh:19]: I'll fix.

Comment (CurtizJ) [reply tests/queries/0_stateless/02247_written_bytes_quota.sh:19]: Ah, ok, I understood. Now the size in bytes in block is used, which is slightly larger than number of bytes in string.

Comment (vitlibar) [reply tests/queries/0_stateless/02247_written_bytes_quota.sh:19]: Now the value 'qwqw' means 13 bytes will be written no matter if it's `async_insert` or not.
Earlier the quota was used differently - 13 bytes for non-async insert of `qwqw` and 8 bytes for async insert of `qwqw`. Of course both inserts write the same blocks of data, it's just only the quota calculation which changed.

**Pull request merged**